db/db_impl_compaction_flush.cc

```diff
@@ -287,6 +287,27 @@ Status DBImpl::CompactRange(const CompactRangeOptions& options,
   auto cfd = cfh->cfd();
   bool exclusive = options.exclusive_manual_compaction;
 
+  if (!options.allow_write_stall) {
+    InstrumentedMutexLock l(&mutex_);
+    WriteStallCondition write_stall_condition = WriteStallCondition::kNormal;
+    do {
+      if (write_stall_condition != WriteStallCondition::kNormal) {
+        TEST_SYNC_POINT("DBImpl::CompactRange:StallWait");
+        bg_cv_.Wait();
+      }
+      const auto& mutable_cf_options = *cfd->GetLatestMutableCFOptions();
+      const auto* vstorage = cfd->current()->storage_info();
+      // check whether one extra immutable memtable or an extra L0 file would
+      // cause write stalling mode to be entered. It could still enter stall
+      // mode due to pending compaction bytes, but that's less common
+      write_stall_condition =
+          ColumnFamilyData::GetWriteStallConditionAndCause(
+              cfd->imm()->NumNotFlushed() + 1,
+              vstorage->l0_delay_trigger_count() + 1,
```

Comment:
How about +2?
Also is it possible that l0 slowdown trigger threshold is so small that there is a chance that we'll never flush here?

Maybe we should skip the check if the threshold is too small.

Reply:
threshold is too small means delay trigger is similar to compaction trigger? I was thinking if they're equal this code will have some problem.

Reply:
idk about + 2, adding more heuristics gives me a headache. I prefer + 1 since we're only flushing one file, unless it causes a problem

Reply:
The case of small slowdown triggers is handled on 312 - 317. Note that `SanitizeOptions` enforces compaction stall trigger is at least `level0_file_num_compaction_trigger`, and flush stall trigger is at least `min_write_buffer_number_to_merge`.

Reply:
But if an organic flush happens just before the manual flush, we end up with flushing two times together and stall is triggered.

Reply:
That's true. Let me land this and think about it more. It is probably safe.